src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComClassGenerator.cs

```diff
@@ -0,0 +1,231 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Microsoft.Interop
+{
+    [Generator]
+    public class ComClassGenerator : IIncrementalGenerator
+    {
+        private sealed record ComClassInfo(string ClassName, ContainingSyntaxContext ContainingSyntaxContext, ContainingSyntax ClassSyntax, SequenceEqualImmutableArray<string> ImplementedInterfacesNames);
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            // Get all types with the [GeneratedComClassAttribute] attribute.
+            var attributedClasses = context.SyntaxProvider
+                .ForAttributeWithMetadataName(
+                    TypeNames.GeneratedComClassAttribute,
+                    static (node, ct) => node is ClassDeclarationSyntax,
+                    static (context, ct) =>
+                    {
+                        var type = (INamedTypeSymbol)context.TargetSymbol;
+                        var syntax = (ClassDeclarationSyntax)context.TargetNode;
+                        ImmutableArray<string>.Builder names = ImmutableArray.CreateBuilder<string>();
+                        foreach (INamedTypeSymbol iface in type.AllInterfaces)
+                        {
+                            if (iface.GetAttributes().Any(attr => attr.AttributeClass?.ToDisplayString() == TypeNames.GeneratedComInterfaceAttribute))
+                            {
+                                names.Add(iface.ToDisplayString());
+                            }
+                        }
+                        return new ComClassInfo(
```

Comment:
Do we want to warn / bail if there are no interfaces with `GeneratedComInterface`?

Reply:
I think that's still a valid scenario, but I could see us adding a warning for it. I'll file a follow-up issue for that.